ffi/src/ATen/GC.hs

```diff
@@ -0,0 +1,56 @@
+{-# LANGUAGE EmptyDataDecls #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE ForeignFunctionInterface #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+
+module ATen.GC where
+
+import Control.Exception.Safe (catch,throwIO)
+import Data.List (isPrefixOf)
+import Language.C.Inline.Cpp.Exceptions (CppException(..))
+import System.Mem (performGC)
+import Control.Concurrent (threadDelay)
+import Control.Concurrent.Async
+import System.SysInfo
+
+retryWithGC' :: Int -> IO a -> IO a
+retryWithGC' count func =
+  func `catch` \a@(CppStdException message) ->
+    if isPrefixOf msgOutOfMemory message
+    then
+      if count <= 0
+      then throwIO $ CppStdException $ "Too many calls to performGC, " ++ message
+      else do
+        performGC
+        threadDelay 1000 -- We need delta delay(1ms) to wait GC.
+        retryWithGC' (count-1) func
+    else throwIO a
+  where
+    msgOutOfMemory :: String
+    msgOutOfMemory = "Exception: CUDA out of memory."
+
+retryWithGC :: IO a -> IO a
+retryWithGC = retryWithGC' 10
+
+checkOSMemoryWithGC :: IO ()
```

Comment:
Well this only pools the CPU memory, not the GPU. The c10 library exposes the current GPU memory usage `c10::cuda::CUDACachingAllocator::currentMemoryAllocated(device)`, but you'll need to get the total amount from somewhere else (or send a patch to add this to upstream, because it's useful).